patches/server/1048-Item-serialization-as-json.patch

```diff
@@ -0,0 +1,88 @@
++        .xmap(com.mojang.datafixers.util.Either::unwrap, data -> { // Both will be used for deserialization, but we decide which one to use for serialization
++            if (!SERIALIZE_CUSTOM_AS_SNBT.get()) {
++                return com.mojang.datafixers.util.Either.left(data); // First codec
++            } else {
++                return com.mojang.datafixers.util.Either.right(data); // Second codec
++            }
++        })
++        // Paper end - Item serialization as json
+         .xmap(CustomData::new, component -> component.tag);
+     public static final Codec<CustomData> CODEC_WITH_ID = CODEC.validate(
+         component -> component.getUnsafe().contains("id", 8) ? DataResult.success(component) : DataResult.error(() -> "Missing id for entity in: " + component)
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+index 317af98f8bd166a5a534f74e6ba177be9afb6a8f..264ff92c35ca68464e580da65a1324fb14ece159 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+@@ -164,6 +164,36 @@ public final class CraftItemFactory implements ItemFactory {
+         }
+     }
+ 
++    // Paper start - Item serialization to json
++    @Override
++    public ItemStack createItemStack(com.google.gson.JsonObject data) throws IllegalArgumentException {
++        Preconditions.checkNotNull(data, "null cannot be deserialized");
++
++        final int dataVersion = data.get("DataVersion").getAsInt();
++        final int currentVersion = org.bukkit.craftbukkit.util.CraftMagicNumbers.INSTANCE.getDataVersion();
++        data = ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertJson(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.ITEM_STACK, data, false, dataVersion, currentVersion);
++        com.mojang.serialization.DynamicOps<com.google.gson.JsonElement> ops = MinecraftServer.getServer().registryAccess().createSerializationContext(com.mojang.serialization.JsonOps.INSTANCE);
++        return CraftItemStack.asCraftMirror(net.minecraft.world.item.ItemStack.CODEC.parse(ops, data).getOrThrow(IllegalArgumentException::new));
++    }
++
++    public static com.google.gson.JsonObject serializeAsJson(net.minecraft.world.item.ItemStack handle) {
++        Preconditions.checkNotNull(handle, "Cannot serialize empty ItemStack");
++        Preconditions.checkArgument(!handle.isEmpty(), "Cannot serialize empty ItemStack");
++
++        net.minecraft.core.RegistryAccess.Frozen reg = net.minecraft.server.MinecraftServer.getServer().registryAccess();
++        com.mojang.serialization.DynamicOps<com.google.gson.JsonElement> ops = reg.createSerializationContext(com.mojang.serialization.JsonOps.INSTANCE);
++        com.google.gson.JsonObject item;
++        net.minecraft.world.item.component.CustomData.SERIALIZE_CUSTOM_AS_SNBT.set(true);
```

Comment:
I assume this is done to properly retain types? if so, you should add a comment explaining that here